### ISSUE TITLE
How to use compute_test_results.m script

### ISSUE THREAD

**Issue by trminh89** [alexgkendall/SegNet-Tutorial#7]
Hi @alexgkendall ,

Thank for uploading the evaluation script. However, could you please write a few lines about how to use this script with your CamVid dataset? I don't familiar with Matlab thought.

Thanks

**Comment by Vaan5** [comment]
Just replace the two "??" symbols with lines that load your ground truth and prediction images e.g.
```
% the same code as in the repo -> if you use another test/train set change number of classes and the
% unlabeled index as well as number of iterations (needs to be equal to the test set size)

gtPath = '/SegNet/Scripts/gt'; % path to your ground truth images
predPath = '/SegNet/Scripts/predictions'; %path to your predictions (you get them after you implement saving images in the test_segmentation_camvid.py script - or you write your own)
groundTruths = dir(gtPath);
skip = 2; % first two are '.' and '..' so skip them
predictions = dir(predPath);

for i = 1:iter
    display(num2str(i));
      
    pred = imread(strcat(predPath, '/', predictions(i + skip).name)); % set this to iterate through your segnet prediction images
    pred = pred + 1; % i added this cause i labeled my classes from 0 to 11
    annot = imread(strcat(gtPath, '/', groundTruths(i + skip).name)); % set this to iterate through your ground truth annotations
    annot = annot + 1; % i added this cause i labeled my classes from 0 to 11 -> so in that case the next line will find every pixel labeled with unknown_class=12
    
    pixels_ignore = annot == unknown_class;
    pred(pixels_ignore) = 0;
    annot(pixels_ignore) = 0;

% the rest of the code

```

**Comment by trminh89** [comment]
Thanks @Vaan5 

I checked the test_segmentation_camvid.py script, and the prediction images was saved using:
      #scipy.misc.toimage(rgb, cmin=0.0, cmax=255).save(IMAGE_FILE+'_segnet.png')

That means this image has 3 channels RGB, and each pixel doesn't store the label, it stores the color instead.

Is it ok if I use this image as the input for evaluation script in Matlab?

**Comment by Vaan5** [comment]
You should change that so that it saves a grayscale image (just like the ones in the testannot dataset)

**Comment by trminh89** [comment]
Thanks, I manage to save the prediction image as in the testannot folder. 

However, I'm trying to save its label from the net. Does the net load label from this command?
        label = net.blobs['label'].data

I check this array and all values are  0, which not correct. How do we save back the label (or know which input file is being processed)?

**Comment by trminh89** [comment]
Hi @alexgkendall and @Vaan5 

How did you save the grayscale image with Python? My labels from label = net.blobs['label'].data are all zero. I check the label image with Matlab and it's ok. The only different between your label in testandnot folder and my label is yours has uint8 type, while my label image has uint16 type. Is it a problem?

Thanks!

**Comment by Vaan5** [comment]
You can use Image from PIL e.g.
The labels can be gotten from the net after a forward pass
```
# i have them as np.uint8 typed arrays
img = Image.fromarray(labels, 'L')
img.save('gt/test' + str(s) + '.ppm') %or another format
```

**Comment by alexgkendall** [comment]
Thank's @Vaan5 I've added your contributions to the test script.

@trminh89 , I'm not sure how the dense image data loading layer in SegNet will handle int32 PNG images. If it doesn't work then you could try encoding your data into an lmdb file (as a float) and using caffe's standard data layer?

**Comment by trminh89** [comment]
Thanks @Vaan5 @alexgkendall !
The problem is the datatype. I managed to write the data as unit8 and it's ok now. 

I don't understand the following two lines in the evaluation script:
    pred = pred + 1; % i added this cause i labeled my classes from 0 to 11
    annot = annot + 1; % i added this cause i labeled my classes from 0 to 11 -> so in that case the next line will find every pixel labeled with unknown_class=12

Could you please explain it more details? Is this script are used for CamVid dataset? I saw in segnet_basic_train.prototxt file that the loss layer ignore label 11 (ignore_label: 11), and it's not very clear about which class is being ignored.

In my cases, my labels are: 0 1 2 3 4 5, while 0 is the background and I want to exclude it from the evaluation. How should I do it in both the segnet_basic_train.prototxt file and the evaluation script?

**Comment by alexgkendall** [comment]
As it is currently written, the Matlab script iterates through classes starting with 1. The caffe model is designed for class indexes counting from 0. So you should add 1 to your ground truth and prediction index maps and unknown index in the Matlab script, which is the purpose of these lines.

You may find it helpful to display both segmentation images in matlab to ensure that the class IDs match.

**Comment by trminh89** [comment]
Thanks @alexgkendall !
I got the index now. 

In the segnet_basic_train.prototxt file, class with label = 11 was ignored. So, when we do the training, caffe won't count this class accuracy. Am I right? As the consequence, when evaluate the test data, should I "ignore" this class too? I don't understand the meaning of ignore_label in training step and how to deal with this class in testing step. Many thanks!

**Comment by alexgkendall** [comment]
Yes use the ignore label for unlabelled or unknown classes.

**Comment by Nirmitee** [comment]
when I am running the test_segmentation_camvid.py script only one  prediction image is getting saved. I am new to python and not understanding why this is happening. Also when I executed compute_test_results.m I get an error saying 
Index exceeds matrix dimensions.
Error in compute_test_results (line 22)
    pred = imread(strcat(predPath,
    '/SegNet/SegNet-Tutorial/Scripts/test.txt+_segnet.png',
    predictions(i + skip).name)); % set this to iterate
    through your segnet prediction images 

This is probably because my prediction images are not getting saved in the right manner. Please can anyone help me?
Thanks in advance

**Comment by HaoWumarz** [comment]
Hi @trminh89 
How did u save the prediction. Could u show an example?
Like how to define the IMAGE_FILE
Thank you
